File: common/src/main/java/com/ishow/common/widget/watermark/WaterMarkHelp.java
package com.ishow.common.widget.watermark;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import android.support.annotation.NonNull;
import android.text.TextUtils;
import android.view.View;

import com.ishow.common.R;

/**
 * Created by yuhaiyang on 2018/1/23.
 * 水印的参数
 */

public class WaterMarkHelp {
    /**
     * 释放一个全局控制的变量
     * 方便以后进行宏控制
     */
    private static boolean sShowWaterMark = true;
    private static String sDefaultText = "测试版本";
    private Paint mPaint;
    private Params mParams;
    private View mView;
    private int mTextWidth;
    private int mTextHeight;

    public void init(View view, @NonNull Params params) {
        mParams = params;
        mView = view;

        params.alpha = Math.min(1.0F, Math.max(0F, params.alpha));
        params.angle = Math.min(360, Math.max(0, params.angle));
        params.text = TextUtils.isEmpty(params.text) ? sDefaultText : params.text;

        mPaint = new Paint(Paint.ANTI_ALIAS_FLAG | Paint.DITHER_FLAG);
        mPaint.setTextSize(params.textSize);
        mPaint.setColor(params.textColor);
        mPaint.setAlpha((int) (params.alpha * 255));

        Rect rect = new Rect();
        mPaint.getTextBounds(params.text, 0, params.text.length(), rect);
        mTextWidth = rect.width() + params.startPadding + params.endPadding;
        mTextHeight = rect.height() + params.topPadding + params.bottomPadding;
    }

    public void draw(Canvas canvas, final int width, final int height) {
        if (!sShowWaterMark) {
            return;
        }

        if (mParams == null || mPaint == null || canvas == null) {
            return;
        }

        if (!mParams.enable) {
            return;
        }

        int count = canvas.saveLayer(0, 0, width, height, mPaint, Canvas.ALL_SAVE_FLAG);
        canvas.rotate(mParams.angle, width / 2, height / 2);

        final int startY = -height;
        final int endY = height + height;
        final int startX = -width;
        final int endX = width + width;

        for (int y = startY; y <= endY; y += mTextHeight) {
            for (int x = startX; x <= endX; x += mTextWidth) {
                canvas.drawText(mParams.text, x + mParams.startPadding, y + mParams.topPadding, mPaint);
            }
        }

        canvas.restoreToCount(count);
    }


    @SuppressWarnings("WeakerAccess")
    public void setWaterMarkEnable(boolean enable) {
        mParams.enable = enable;
        mView.postInvalidate();
    }

    @SuppressWarnings("WeakerAccess")
    public static int getDefaultTextSize(Context context) {
        return context.getResources().getDimensionPixelSize(R.dimen.A_title);
    }

    @SuppressWarnings("WeakerAccess")
    public static int getDefaultTextColor(Context context) {
        return context.getResources().getColor(R.color.text_grey_light_normal);
    }

    @SuppressWarnings("WeakerAccess")
    public static int getDefaultPadding(Context context) {
        return context.getResources().getDimensionPixelSize(R.dimen.dp_20);
    }

    @SuppressWarnings("WeakerAccess")
    public static float getDefaultAlpha() {
        return 0.2F;
    }

    public static int getDefaultAngle() {
        return 315;
    }

    public static void show(boolean show) {
        sShowWaterMark = show;
    }

    public static void defaultText(String text) {
        if (TextUtils.isEmpty(text)) {
            return;
        }
        sDefaultText = text;
    }

    public static class Params {
        public String text;
        public int textSize;
        public int textColor;
        public int topPadding;
        public int bottomPadding;
        public int startPadding;
        public int endPadding;
        public float alpha;
        public int angle;
        public boolean enable;
    }
}

File: common/src/main/java/com/ishow/common/widget/watermark/WaterMarkHelp.kt
package com.ishow.common.widget.watermark

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Rect
import android.text.TextUtils
import android.view.View
import androidx.core.content.ContextCompat
import com.ishow.common.R
import com.ishow.common.extensions.dp2px
import kotlin.math.max
import kotlin.math.min

/**
 * Created by yuhaiyang on 2018/1/23.
 * 水印的参数
 */

class WaterMarkHelp {
    private lateinit var mPaint: Paint
    private lateinit var mParams: Params
    private var mView: View? = null
    private var mTextWidth: Int = 0
    private var mTextHeight: Int = 0

    fun init(view: View, params: Params) {
        mParams = params
        mView = view

        params.alpha = min(1.0f, max(0f, params.alpha))
        params.angle = min(360, max(0, params.angle))
        params.text = if (params.text.isNullOrBlank()) sDefaultText else params.text

        mPaint = Paint(Paint.ANTI_ALIAS_FLAG or Paint.DITHER_FLAG)
        mPaint.textSize = params.textSize.toFloat()
        mPaint.color = params.textColor
        mPaint.alpha = (params.alpha * 255).toInt()

        val rect = Rect()
        mPaint.getTextBounds(params.text, 0, params.text!!.length, rect)
        mTextWidth = rect.width() + params.startPadding + params.endPadding
        mTextHeight = rect.height() + params.topPadding + params.bottomPadding
    }

    fun draw(canvas: Canvas?, width: Int, height: Int) {
        if (!sShowWaterMark) {
            return
        }

        if ( canvas == null) {
            return
        }

        if (!mParams.enable) {
            return
        }

        val count = canvas.saveLayer(0f, 0f, width.toFloat(), height.toFloat(), mPaint)
        canvas.rotate(mParams.angle.toFloat(), width / 2F, height / 2F)

        val startY = -height
        val endY = height + height
        val startX = -width
        val endX = width + width

        var y = startY
        while (y <= endY) {
            var x = startX
            while (x <= endX) {
                canvas.drawText(mParams.text!!, (x + mParams.startPadding).toFloat(), (y + mParams.topPadding).toFloat(), mPaint)
                x += mTextWidth
            }
            y += mTextHeight
        }

        canvas.restoreToCount(count)
    }


    fun setWaterMarkEnable(enable: Boolean) {
        mParams.enable = enable
        mView?.postInvalidate()
    }

    class Params {
        var text: String? = null
        var textSize: Int = 0
        var textColor: Int = 0
        var topPadding: Int = 0
        var bottomPadding: Int = 0
        var startPadding: Int = 0
        var endPadding: Int = 0
        var alpha: Float = 0.toFloat()
        var angle: Int = 0
        var enable: Boolean = false
    }

    companion object {
        /**
         * 释放一个全局控制的变量
         * 方便以后进行宏控制
         */
        private var sShowWaterMark = true
        private var sDefaultText = "测试版本"

        fun getDefaultTextSize(context: Context): Int {
            return context.resources.getDimensionPixelSize(R.dimen.A_title)
        }

        fun getDefaultTextColor(context: Context): Int {
            return ContextCompat.getColor(context, R.color.text_grey_light)
        }

        fun getDefaultPadding(context: Context): Int {
            return 20.dp2px()
        }

        val defaultAlpha: Float
            get() = 0.2f

        val defaultAngle: Int
            get() = 315

        fun show(show: Boolean) {
            sShowWaterMark = show
        }

        fun defaultText(text: String) {
            if (!TextUtils.isEmpty(text)) {
                sDefaultText = text
            }
        }
    }
}
